File: test/scripts/string_test.py
#
# This file is part of libdestruct (https://github.com/mrindeciso/libdestruct).
# Copyright (c) 2024 Roberto Alessandro Bertolini. All rights reserved.
# Licensed under the MIT license. See LICENSE file in the project root for details.
#

import unittest

from libdebug import debugger
from libdestruct import inflater, c_str

class StringTest(unittest.TestCase):
    def test_basic(self):
        d = debugger("binaries/string_test")

        def callback(_, __):
            pass

        d.run()

        bp1 = d.bp("signpost1")
        bp2 = d.bp("signpost2")
        check = d.bp("check", callback=callback)

        d.cont()

        assert bp1.hit_on(d)

        libdestruct = inflater(d.memory)

        string = libdestruct.inflate(c_str, d.regs.rdi)

        self.assertEqual(string.value, b"Hello, world!")

        string.value = b"Hello, WORLD!"

        d.cont()

        assert bp2.hit_on(d)

        self.assertEqual(string.value, b"Hello,")

        string.value = b", world!\x00"

        d.cont()

        d.kill()

        assert check.hit_count == 0

        d.terminate()

File: test/srcs/string_test.c
//
// This file is part of libdestruct (https://github.com/mrindeciso/libdestruct).
// Copyright (c) 2024 Roberto Alessandro Bertolini. All rights reserved.
// Licensed under the MIT license. See LICENSE file in the project root for details.
//

// gcc -o binaries/string_test srcs/string_test.c

#include <stdio.h>
#include <stdlib.h>
#include <string.h>
#include <sys/mman.h>

void signpost1(char *address)
{

}

void signpost2()
{

}

void check()
{

}

int main()
{
    char test_string[] = "Hello, world!\0";

    signpost1(test_string);

    if (strcmp(test_string, "Hello, WORLD!") != 0) {
        check();
    }

    test_string[6] = '\0';

    signpost2();

    if (strcmp(test_string, ", world!") != 0) {
        check();
    }

    return 0;
}